contents/blog/how-we-built-an-app-server.md

```diff
@@ -0,0 +1,304 @@
+
+Like many features, building this started as a hackathon MVP and developed into a service that now handles billions of events. Here are how we built it and the lessons we learned along the way.
+
+## The three-day app MVP
+
+Apps started as an MVP that came together in 3 days. Marius created it to scratch his itch for a world map that existed in Google Analytics but didn’t in PostHog. The map required location data from IP addresses. There were many services to do this, but connecting them to event data going was difficult.
+
+![Map](../images/blog/how-we-built-an-app-server/map.png)
+
+The idea expanded to enable more modification of the events pipeline such as backing up data to S3, syncing GitHub stars, or getting customer feedback.
+
+### Arbitrary app code in Python
+
+The three-day hackathon led to the ability to write and run arbitrary Python code. It got downloaded from GitHub, extracted, and integrated into PostHog. It specifically required Python 3.9, because Python 3.8’s `zipimport` didn’t support the zip archives GitHub provides.
+
+An app at this point looked like this:
+
+```python
+# exampleplugin/__init__.py
+from posthog.plugins import PluginBaseClass, PosthogEvent, TeamPlugin
+class ExamplePlugin(PluginBaseClass):   
+    def __init__(self, team_plugin_config: TeamPlugin):
+        super().__init__(team_plugin_config)
+        # other per-team init code
+    def process_event(self, event: PosthogEvent):
+        event.properties["hello"] = "world"
+        return event
+    def process_identify(self, event: PosthogEvent):
+        pass
+```
+
+Other features added included a plugin repository, an interface to configure the apps, and a CLI to preconfigure apps for custom installations. Marius built two sample plugins with it: the currency normalizer (convert event properties to a single currency), and the coveted GeoIP app.
+
+It didn’t take long for cracks to appear. First, it’s nearly impossible to manage Python dependencies via multiple `requirements.txt` files. There's no way to tell which dependencies your plugin's dependencies will install, without first installing them. There is [nothing like `pip --dry-run`](https://github.com/pypa/pip/issues/53). You need to install the `pip` package and run an `__init__.py` script inside the package to get its dependency tree.
+
+What's more, all dependencies get installed together. This makes it just a matter of time before a plugin overrides an app dependency (e.g. `django`), and brings everything crashing down. This wasn’t going to work for us, so we needed to rebuild. 
+
+## Rebuilding in JavaScript
+
+After deciding to rebuild, the first attempt was using [PyMiniRacer](https://github.com/sqreen/PyMiniRacer) to generate JavaScript. PyMiniRacer is great for simple functions, but lacked support for async/promises and importing modules like `fetch` because it was raw v8, not Node. It also had a limited standard library. All of which led to the decision it also wasn’t going to work. 
```

Comment:
```suggestion
After deciding to rebuild, the first attempt was using [PyMiniRacer](https://github.com/sqreen/PyMiniRacer) to generate JavaScript. PyMiniRacer is great for simple functions, but lacks support for async/promises and importing modules like `fetch` because it was raw v8, not Node. It also had a limited standard library. It wasn’t going to work. 
```